FEAT Allow appending php version to username folder

lekoala opened the pull request:
## Description
If you want to link the php version to the temp folder, this is currently only possible using the system path.
This changes allows an optional env variable that will append the php version to the username folder
With this, you can use the local silverstripe-cache folder and still get per php version cache folders

I'm not sure why the php version is added by default in the system path. If this is a desired behaviour to link the php version to the cache folder, maybe this should be the default also for the local cache?
If it's not necessary, it could be more consistent to add the php version to the username folder instead.

## Manual testing steps
- Enable the TEMP_FOLDER_VERSION=true
- Create a local silverstripe-cache folder
- See that the username now contains the php version

## Issues
https://github.com/silverstripe/silverstripe-framework/issues/11745

## Pull request checklist
<!--
  PLEASE check each of these to ensure you have done everything you need to do!
  If there's something in this list you need help with, please ask so that we can assist you.
-->
- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [ ] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [ ] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

lekoala committed:
FEAT Allow appending php version to username folder

GuySartorelli commented on src/Core/TempFolder.php:
```diff
@@ -18,9 +18,15 @@ class TempFolder
+        $user = static::getTempFolderUsername();
+
+        // Append php version to username folder
+        if (Environment::getEnv('TEMP_FOLDER_VERSION')) {
```
```suggestion
        if (Environment::getEnv('SS_TEMP_FOLDER_INCLUDES_PHP_VERSION')) {
```
Just to make it clearer what this controls and add the standard `SS_` prefix.

Should maybe even be `SS_PROJECT_TEMP_FOLDER_INCLUDES_PHP_VERSION` since it doesn't affect the temp folder if it's in tmp/... but that's pretty long so probably we can let documentation cover that detail.

GuySartorelli commented on src/Core/TempFolder.php:
```diff
@@ -18,9 +18,15 @@ class TempFolder
     public static function getTempFolder($base)
     {
         $parent = static::getTempParentFolder($base);
+        $user = static::getTempFolderUsername();
```
```suggestion
        $folderName = static::getTempFolderUsername();
```
It's not just the user anymore

GuySartorelli commented on src/Core/TempFolder.php:
```diff
@@ -18,9 +18,15 @@ class TempFolder
+            $user .= '-' . preg_replace('/[^\w\-\.+]+/', '-', PHP_VERSION);
+        }
 
         // The actual temp folder is a subfolder of getTempParentFolder(), named by username
```
```suggestion
        // The actual temp folder is a subfolder of getTempParentFolder(), named by username and optionally including PHP version
```

GuySartorelli reviewed (changes requested):
Thanks for raising this. Please:

1. Retarget this PR to the `6` branch. There will be no more minor releases for CMS 5.
2. Change `FEAT` for `ENH` in the commit message as per the commit message guidelines
3. create a PR to update the docs:
    - Add a note about this to the changelog (probably just a bullet point in the "Other new features" section)
    - Update the table in https://docs.silverstripe.org/en/6/getting_started/environment_management/#core-environment-variables

lekoala committed:
Update src/Core/TempFolder.php

Co-authored-by: Guy Sartorelli <36352093+GuySartorelli@users.noreply.github.com>

lekoala committed:
Update src/Core/TempFolder.php

Co-authored-by: Guy Sartorelli <36352093+GuySartorelli@users.noreply.github.com>

lekoala committed:
Update src/Core/TempFolder.php

Co-authored-by: Guy Sartorelli <36352093+GuySartorelli@users.noreply.github.com>

lekoala commented:
@GuySartorelli what do you think about the difference between the system temp folder (that is always using the php version) et the local temp folder ? What is the reason behind always having the php version in the system temp folder ?

Also, I'm thinking about 
https://github.com/silverstripe/silverstripe-framework/issues/11657

Should it take care of cleaning "old" temp folders from previous php versions ?

lozcalver commented:
> What is the reason behind always having the php version in the system temp folder?

From memory this originated from issues with serialization - different PHP versions would serialize things slightly differently, so if you changed PHP version then things would break because the cached files would fail to decode.

Edit: found it https://github.com/silverstripe/silverstripe-framework/pull/6810

lekoala commented:
yes, exactly. in that case, wouldn't it make sense to make that a default instead of an opt in setting for local cache as well?

GuySartorelli commented:
> Also, I'm thinking about
> https://github.com/silverstripe/silverstripe-framework/issues/11657
> Should it take care of cleaning "old" temp folders from previous php versions ?

Probably worth discussing that _in that issue_ - but I would expect it to only touch cache that is either managed by a cache adapter or contained inside `TEMP_PATH`.

GuySartorelli commented:
> @GuySartorelli what do you think about the difference between the system temp folder (that is always using the php version) et the local temp folder ? What is the reason behind always having the php version in the system temp folder?

Loz beat me to finding the answer to that (which is good because I don't have time to dive into historical details for this right now).
Looks like #6810 was meant to solve this for the `silverstripe-cache/` dir in the project root, so I'm curious as to why that was reverted. If you have time to discover that, and can show that either there was no clear reason or that it was a mistake, then we can do away with the environment variable.

But without knowing why the php version was removed from the folder name after that PR I'd be hesitant to just blindly add it again without it being opt-in

(Also the CI failure looks related to this PR, can you please check that out?)

lekoala committed:
Update TempFolder.php

lekoala commented:
@GuySartorelli ah yes fixed the CI failure, it was due to the renaming of the variable

the change for the php version was introduced in the v3
https://github.com/silverstripe/silverstripe-framework/blob/3.7/core/TempPath.php
the logic there was to have a version neutral parent folder, and appending the php version to the subfolder (for system and local path)

Back then, there was no "TempFolder" class. In v4, with the new class, it seems the logic has changed. Maybe the work on the v4 had already started, the TempFolder class was created based on an older v3 logic (last change 8 years ago), and the changes never made it upstream (last change in v3 is 7 years ago). That would be my guess.

I think my issue is consistent with https://github.com/silverstripe/silverstripe-framework/issues/2986

Therefore, I suggest to reapply the same logic:
- a version neutral parent folder (silverstripe-cache or silverstripe-cache-$base)
- always append the php version to the subfolder
- as an additional feature, we could add an env variable to disable the php version so that you can have the old behaviour for local folder, or a version neutral system folder (new feature)

GuySartorelli commented:
> a version neutral parent folder (silverstripe-cache or silverstripe-cache-$base)

I assume no change is necessary for this part?

> always append the php version to the subfolder

Sounds sensible to me

> as an additional feature, we could add an env variable to disable the php version so that you can have the old behaviour for local folder, or a version neutral system folder (new feature)

I don't think that's necessary. It sounds like there isn't likely to be a good reason to have the cache shared between PHP versions. If someone finds such a reason, we can always add an environment variable to toggle it after the fact.

lekoala committed:
always append php version to username folder

lekoala committed:
Update and rename CoreTest.php to TempFolderTest.php

lekoala commented:
@GuySartorelli i've updated the PR:
- Renamed CoreTest to TempFolderTest because that's what it is
- Always test all cases (local and system folders)
- Removed ?? '' that don't bring any benefit (the Path:join always returns a string)
- Always append the php version to the username folder and have a version neutral parent folder in the system temp dir

let me know what you think :) i hope it's not an issue to do cleanup like removing ?? '', if needed I can put them back but I don't really see the point and it makes the code less readable. Static analysis should still be happy due to return types.

michalkleiner commented:
Just passing by, but wanted to say thanks @lekoala for the contributions you're working on, it's a nice help and good additions/enhancements. We do appreciate it!

GuySartorelli reviewed (approved):
Sorry it's taken me so long to look at this again.
I've rebased the PR on the latest changes in the `6` branch.
It works well locally. If CI goes green, I'll merge.

Can you please submit a PR to developer-docs adding a note to the 6.1.0 changelog? Just a short bullet point under "other new features and enhancements" would be fine. It should note the following:
1. Inside your cache directory, sub-directories now include the PHP version, e.g. `my-user-8.3.23/`. Previously the PHP version was not included.
1. If you do not have a `silverstripe-cache/` directory in your project root, the parent cache directory that gets created no longer contains the PHP version (e.g. it will be `/tmp/silverstripe-cache-var-www-html/` where it used to be `/tmp/silverstripe-cache-php8.3.23-var-www-html/`).

lekoala commented:
@GuySartorelli no worries, i've created the PR for the docs

GuySartorelli commented:
Sweet, thank you.
I'm just holding off for a beat before merging this - I had a realisation that I'm not sure how cache is being shared in Silverstripe Cloud hosting so I'm getting someone from that team across this in case they're expecting e.g. a folder that matches `/tmp/silverstripe-cache-php.*` in order to share cache between servers.
